### ISSUE TITLE
Losses for E2E Training

### ISSUE THREAD

**Issue by zanussbaum** [XiangLi1999/Diffusion-LM#47]
Hi, thanks for releasing the code! I had a quick question about the different loss functions in the code. 

I'm trying to wrap my head around the loss function presented in the paper and compare it to what's in the code. I'm taking a look at the function 
`Le2esimple(w) = Eqφ(x0:T |w) Lsimple(x0) + ||EMB(w) − µθ(x1, 1)||2 − log pθ(w|x0)`

LSimple appears to be [this line](https://github.com/XiangLi1999/Diffusion-LM/blob/352b28c623c83083ab31d0d0b2a611f01c0fcb6c/improved-diffusion/improved_diffusion/gaussian_diffusion.py#L1557)

The loss between the embeddings seem to be [these lines](https://github.com/XiangLi1999/Diffusion-LM/blob/352b28c623c83083ab31d0d0b2a611f01c0fcb6c/improved-diffusion/improved_diffusion/gaussian_diffusion.py#L1561-L1563)

And the cross entropy loss between the logits and input tokens appears to [be here](https://github.com/XiangLi1999/Diffusion-LM/blob/352b28c623c83083ab31d0d0b2a611f01c0fcb6c/improved-diffusion/improved_diffusion/gaussian_diffusion.py#L1569)

However, I'm a little confused on what these [lines account for](https://github.com/XiangLi1999/Diffusion-LM/blob/352b28c623c83083ab31d0d0b2a611f01c0fcb6c/improved-diffusion/improved_diffusion/gaussian_diffusion.py#L1567). From my debugging, this just seems to be taking the embeddings multiplied with noise, multiplied with `sqrt_alphas_cumprod` across all timesteps. 

Am I misinterpreting what's in the code versus what's in the paper?

**Comment by XiangLi1999** [comment]
Hi, 

Thanks for the great question! 

We are simplifying a bit in the main text of the paper, and you can regard this term tT_loss as part of Lsimple when t=T. It's a bit misuse of notation to simplify the main text: that by assumption \mu_theta(x_T, T)= 0 since p(x_T) = N(0, I), so this Lsimple term at t=T reduces to a L2 norm of x_T. 

The Lsimple term is derived from Lvlb term, and we put a detailed derivation in appendix E (page 17), 
<img width="664" alt="image" src="https://user-images.githubusercontent.com/29054786/202839183-1b24115a-4bcd-4cbf-81d2-c2a2cbfa085d.png"> and this line in particular justifies the term tT_loss. 

Intuitively, this term exists to avoid the embedding norms from being too large.

**Comment by zanussbaum** [comment]
Thanks for the great explanation! I missed that part originally when looking over the paper but intuitively makes sense to have a L2 norm of x_T!
